### PR TITLE
WP-7832 Import contact list contents from Hubspot.

### PR DESCRIPTION
**Pull request by ichidyakin**
Contact lists are now discovered as streams available for syncing.

# Description of change
JIRA: https://varicent.atlassian.net/browse/WP-7832
This change adds existing contact lists in Hubspot available for syncing as separate streams

# Rollback steps
 - revert this branch

# Tags to create
`v2.9.7`

### CODE REVIEW

**Review by pggagnon — CHANGES_REQUESTED** [review]
Functionally I think everything looks pretty good. A couple of changes though:
- This PR should be targeting the tap-hubspot repo.
- Can we add tests for this new functionality? Did you look into what tests might be appropriate for this addition?

File: tap_hubspot/__init__.py [hunk 1]
```diff
@@ -189,7 +191,8 @@ def load_associated_company_schema():
     return associated_company_schema
 
 def load_schema(entity_name):
-    schema = utils.load_json(get_abs_path('schemas/{}.json'.format(entity_name)))
+    entity_class = entity_name.split(":")[0]
```
Comment (kchan-varicent) [inline tap_hubspot/__init__.py:194]: is this change needed if line 940 does the same thing or does it fix other entities?

File: tap_hubspot/__init__.py [hunk 2]
```diff
@@ -392,7 +395,7 @@ def get_v3_deals(v3_fields, v1_data):
     return v3_resp.json()['results']
 
 #pylint: disable=line-too-long
-def gen_request(STATE, tap_stream_id, url, params, path, more_key, offset_keys, offset_targets, v3_fields=None):
+def gen_request(STATE, tap_stream_id, url, params, path, more_key, offset_keys, offset_targets, v3_fields=None, write_state=True):
```
Comment (kchan-varicent) [inline tap_hubspot/__init__.py:398]: alternative idea: instead of adding parameter, how about updating this method to handle if STATE is None? discover shouldn't have state so it can pass in None and this method can check STATE before getting/setting/writing it

**Review by kchan-varicent — CHANGES_REQUESTED** [review]

File: tap_hubspot/__init__.py [hunk 3]
```diff
@@ -89,9 +89,11 @@ class StateFields:
 
     "engagements_all":        "/engagements/v1/engagements/paged",
 
+    "contact_lists":        "/contacts/v1/lists",
```
Comment (sunil-varicent) [inline tap_hubspot/__init__.py:92]: `contact_lists` and `contact_list` is confusing in what they mean. Could we have some other name for list of persons (contacts) in a given contact list? maybe `contacts`?

File: tap_hubspot/__init__.py [hunk 4]
```diff
@@ -929,6 +934,51 @@ def sync_deal_pipelines(STATE, ctx):
...
+                        row[bookmark_key],
+                        UNIX_MILLISECONDS_INTEGER_DATETIME_PARSING))
+
+            if not modified_time or modified_time >= start:
+                vids.append(row['vid'])
+
+            if modified_time and modified_time >= max_bk_value:
+                max_bk_value = modified_time
+
+            if len(vids) == 100:
```
Comment (sunil-varicent) [inline tap_hubspot/__init__.py:971]: what is special of length 100? inside if block its calling same `_sync_contact_vids` with same param. Then it will again be called outside if block, why do we need to call the method twice?

File: tap_hubspot/__init__.py [hunk 5]
```diff
@@ -1086,6 +1149,8 @@ def main_impl():
     if args.state:
         STATE.update(args.state)
 
+    discover_additional_streams()
```
Comment (sunil-varicent) [inline tap_hubspot/__init__.py:1152]: Maybe directly calling `discover_contact_list_streams` is cleaner

File: tap_hubspot/__init__.py (hunk 3 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:92]: I agree with you, it does look confusing. Sadly, there's an entity called `contacts` already. So I'm open to suggestions

File: tap_hubspot/__init__.py (hunk 1 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:194]: The tricky part is when we discover streams (please, refer to the line 1087). If it was only for loading a schema during the data import I wouldn't implement it here. So there's a decision to be made: either we introduce `stream_class` as an attribute of the `Stream` or we parse `stream_id` here. The latter seems nicer and cleaner to me. Your take?

File: tap_hubspot/__init__.py (hunk 4 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:971]: I took this part from how this `tap` imports contacts. It looks like it's batching requests to the HubSpot server taking contact details of up to 100 contacts at a time.
And after processing all the items in the contact list we call for details of the rest of unsynced contacts

File: tap_hubspot/__init__.py (hunk 5 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:1152]: You might be right about it. I just thought that there might be some additional streams to discover in the future. And also I didn't want to have any specifics regarding business entities in the main function. So I leave this up to you: for me there's no issue in calling it directly.

File: tap_hubspot/schemas/contact_list.json [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        },
+        "company": {
+          "type": "object",
+          "properties": {
+            "value": {
+              "type": "string"
+            }
+          },
+          "required": [
+            "value"
```
[the remark this answers is not in the document]
Comment (ichidyakin) [reply tap_hubspot/schemas/contact_list.json:229]: The trick is that if we have a `company` object in the result it will contain `value`.
Considering that HubSpot may suddenly change their stable API (as they did it once with `filter` property of a contact_lists entity it makes sense to remove these constraints. But from the other hand if they changed their API I'd rather be aware of that and adjust the schema then just let it slide. Your take?

File: tap_hubspot/__init__.py (hunk 2 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:398]: You're absolutely correct that we haven't had `STATE` during dicsovery before this change.
But as for your idea, I'm not sure it's going to work: before this change we didn't issue any requests to the server during discovery. Now we do, and `singer` manages offsets and paginated loading using `STATE` during contact lists load. We're going to need `STATE` to support it, esp. in case when we more than one page of contact lists available.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+
+## 2.9.7
+  * Add support for importing contact list contents [WP-7832](https://varicent.atlassian.net/browse/WP-7832)
```
[the remark this answers is not in the document]
Comment (ichidyakin) [reply CHANGELOG.md]: Thanks, a PR reference comment is addressed.
This version number will definitely be referenced in `app` PR along with a couple more `import-activity` minor changes (version bump etc)

File: tap_hubspot/__init__.py (hunk 2 above)
Comment (kchan-varicent) [reply tap_hubspot/__init__.py:398]: i don't think the singer spec intended state to be used during discovery, i think it's meant to only be used during sync

Comment (ichidyakin) [reply tap_hubspot/__init__.py]: Got it now, thanks! Addressed.

**Comment by ichidyakin** [comment]
@pggagnon ,
> Functionally I think everything looks pretty good. A couple of changes though:
> 
> * ~This PR should be targeting the tap-hubspot repo.~ disregard
> * Can we add tests for this new functionality? Did you look into what tests might be appropriate for this addition?

I don't think there can be any Unit tests added for this change (and they're still green after my changes).
There is a couple of integration tests that might make sense: verify discovery of contact lists and verify data import from them. But there are several obstacles:
- I wasn't able to install `tap_tester` library that's required for testing (and it appears to be an early alpha so to speak).
- Existing integration tests require a live HubSpot account and will make API calls. They also may need some specific entities to exist in HubSpot (or we may need to add some code to create/clean them up on the fly).
- Adding a test will oblige us to actually run it, preferably as a PR Status Check: It seems like a huge amount of work that hasn't been done yet, and at this point I'm too afraid to start it :)

I hope you'll find it reasonable to create a separate ticket for setting up integration tests for this repository.

**Comment by pggagnon** [comment]
@ichidyakin That makes sense to me. I'll leave you to address feedback from other reviews then.

**Review by pggagnon — APPROVED** [review]

File: tap_hubspot/__init__.py (hunk 3 above)
Comment (sunil-varicent) [reply tap_hubspot/__init__.py:92]: `contact_list_entries`?

File: tap_hubspot/__init__.py (hunk 5 above)
Comment (sunil-varicent) [reply tap_hubspot/__init__.py:1154]: I noticed that it runs for discover and sync. Is this necessary to run in sync mode as well?

File: tap_hubspot/schemas/contact_list.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunil-varicent) [reply tap_hubspot/schemas/contact_list.json:229]: Do you mean if Company object is present value param is required, else Company object may be missing, and it runs fine?

[the remark this answers is not in the document]
Comment (ichidyakin) [reply tap_hubspot/schemas/contact_list.json:229]: Yes, exactly

[the remark this answers is not in the document]
Comment (sunil-varicent) [reply tap_hubspot/schemas/contact_list.json:229]: Its good then.

File: tap_hubspot/__init__.py (hunk 3 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:92]: `contact_list_contacts` makes sense to me, thank you!
Addressed

File: tap_hubspot/__init__.py (hunk 5 above)
Comment (ichidyakin) [reply tap_hubspot/__init__.py:1154]: I think it makes sense: we apply user selection to the list of available streams, the code looks nicer. 
Also it seems cheap enough to have it in both places.

Comment (sunil-varicent) [reply tap_hubspot/__init__.py:1154]: is it possible to not discover contact list in sync mode?

Comment (ichidyakin) [reply tap_hubspot/__init__.py:1154]: Yes, it's possible. But there's a downside: the code that decides what streams to sync will become uglier (please refer to `get_streams_to_sync` and `get_selected_streams`) and will contain code specific to handling `contact_list_contacts` entity. Now it's a clean code that relies on the list of streams (`STREAMS` variable) and does not know anything specific to a particular stream

Comment (sunil-varicent) [reply tap_hubspot/__init__.py:1154]: What if the `discover` step outputs necessary contact list details in properties file, and in sync mode it only reads details from properties file and do further steps required for sync? Is this flow possible in this tap?

Comment (ichidyakin) [reply tap_hubspot/__init__.py]: Sunil, thanks again for the detailed explanation. Addressed. Please refer to `populate_streams_from_catalog()` method that populates streams without fetching them from the HubSpot server

File: tap_hubspot/__init__.py [hunk 8]
```diff
@@ -1002,6 +1056,15 @@ def do_sync(STATE, catalog):
...
     LOGGER.info("Sync completed")
 
+def populate_streams_from_catalog(catalog):
+    for stream in catalog.get('streams'):
+        stream_id = stream['tap_stream_id']
+        # Populating contact lists
+        if (stream_id.startswith('contact_list:')):
+            LOGGER.info('Populating contact list stream "%s"', stream_id)
+            stream = Stream(stream_id, sync_contact_list_contacts, ["vid"], 'versionTimestamp', 'FULL_TABLE'),
+            STREAMS.append(stream[0])
```
[the remark this answers is not in the document]
Comment (ichidyakin) [reply tap_hubspot/__init__.py:1066]: Thanks for noticing, it was a typo. Fixed

[the remark this answers is not in the document]
Comment (sunil-varicent) [reply tap_hubspot/__init__.py:1066]: Still missed to change it.

[the remark this answers is not in the document]
Comment (turboslon) [reply tap_hubspot/__init__.py]: Now it's done :)

[the remark this answers is not in the document]
Comment (ichidyakin) [reply tap_hubspot/__init__.py]: Now it's done :)

**Review by sunil-varicent — APPROVED** [review]

**Review by kchan-varicent — CHANGES_REQUESTED** [review]

File: tap_hubspot/__init__.py [hunk 9]
```diff
@@ -396,7 +399,7 @@ def gen_request(STATE, tap_stream_id, url, params, path, more_key, offset_keys,
     if len(offset_keys) != len(offset_targets):
         raise ValueError("Number of offset_keys must match number of offset_targets")
 
-    if singer.get_offset(STATE, tap_stream_id):
+    if STATE != None and singer.get_offset(STATE, tap_stream_id):
```
[the remark this answers is not in the document]
Comment (ichidyakin) [reply tap_hubspot/__init__.py]: Thank you, will use this from now on. Fixed.

**Review by kchan-varicent — APPROVED** [review]
